src/server/notifications/WebHookSubscription2021/WebHookEmitter.ts

```diff
@@ -0,0 +1,90 @@
+import fetch from 'cross-fetch';
+import { calculateJwkThumbprint, importJWK, SignJWT } from 'jose';
+import { v4 } from 'uuid';
+import type { JwkGenerator } from '../../../identity/configuration/JwkGenerator';
+import type { InteractionRoute } from '../../../identity/interaction/routing/InteractionRoute';
+import { getLoggerFor } from '../../../logging/LogUtil';
+import { trimTrailingSlashes } from '../../../util/PathUtil';
+import { readableToString } from '../../../util/StreamUtil';
+import type { NotificationEmitterInput } from '../NotificationEmitter';
+import { NotificationEmitter } from '../NotificationEmitter';
+import type { WebHookFeatures } from './WebHookSubscription2021';
+
+/**
+ * Emits a notification representation using the WebHookSubscription2021 specification.
+ *
+ * Generates a DPoP token and proof, and adds those to the HTTP request that is sent to the target.
+ *
+ * The `expiration` input parameter is how long the generated token should be valid in minutes.
+ * Default is 20.
+ */
+export class WebHookEmitter extends NotificationEmitter<WebHookFeatures> {
+  protected readonly logger = getLoggerFor(this);
+
+  private readonly issuer: string;
+  private readonly webId: string;
+  private readonly jwkGenerator: JwkGenerator;
+  private readonly expiration: number;
+
+  public constructor(baseUrl: string, webIdRoute: InteractionRoute, jwkGenerator: JwkGenerator, expiration = 20) {
+    super();
+    this.issuer = trimTrailingSlashes(baseUrl);
+    this.webId = webIdRoute.getPath();
+    this.jwkGenerator = jwkGenerator;
+    this.expiration = expiration * 60 * 1000;
+  }
+
+  public async handle({ info, representation }: NotificationEmitterInput<WebHookFeatures>): Promise<void> {
+    this.logger.debug(`Emitting WebHook notification with target ${info.features.target}`);
+
+    const privateKey = await this.jwkGenerator.getPrivateKey();
+    const publicKey = await this.jwkGenerator.getPublicKey();
+
+    const privateKeyObject = await importJWK(privateKey);
+
+    // Make sure both header and proof have the same timestamp
+    const time = Date.now();
+
+    // The spec is not completely clear on which fields actually need to be present in the token,
```

Comment:
Have we reported this?

Reply:
No. Issue is also that the webhook "spec" currently is a markdown document that is still much in flux and is not and does not even have a published version yet: https://github.com/solid/notifications/blob/main/webhook-subscription-2021.md